pkg/skaffold/kubernetes/portforward/kubectl_forwarder.go

```diff
@@ -20,105 +20,144 @@ import (
-				ln.Close()
-				return false, nil
+		pfe.cancel = cancel
+
+		cmd := k.kubectl.Command(ctx,
+			"port-forward",
+			"--pod-running-timeout", "1s",
+			fmt.Sprintf("%s/%s", pfe.resource.Type, pfe.resource.Name),
+			fmt.Sprintf("%d:%d", pfe.localPort, pfe.resource.Port),
+			"--namespace", pfe.resource.Namespace,
+		)
+		buf := &bytes.Buffer{}
+		cmd.Stdout = buf
+		cmd.Stderr = buf
+
+		if err := cmd.Start(); err != nil {
+			if ctx.Err() == context.Canceled {
+				logrus.Debugf("couldn't start %v due to context cancellation", pfe)
+				return
 			}
-			return true, nil
-		})
-		resultChan <- err
-	}()
+			//retry on exit at Start()
+			logrus.Debugf("error starting port forwarding %v: %s, output: %s", pfe, err, buf.String())
+			time.Sleep(500 * time.Millisecond)
+			continue
+		}
 
-	err := <-resultChan
-	return err
+		//kill kubectl on port forwarding error logs
+		go k.monitorErrorLogs(ctx, buf, cmd, pfe)
+
+		if err := cmd.Wait(); err != nil {
+			if ctx.Err() == context.Canceled {
+				logrus.Debugf("terminated %v due to context cancellation", pfe)
+				return
+			}
+			logrus.Debugf("port forwarding %v got terminated: %s, output: %s", pfe, err, buf.String())
+			time.Sleep(500 * time.Millisecond)
+		}
+	}
 }
 
 // Terminate terminates an existing kubectl port-forward command using SIGTERM
 func (*KubectlForwarder) Terminate(p *portForwardEntry) {
 	logrus.Debugf("Terminating port-forward %v", p)
 
+	p.terminationLock.Lock()
+	defer p.terminationLock.Unlock()
+
 	if p.cancel != nil {
 		p.cancel()
 	}
+	p.terminated = true
 }
 
 // Monitor monitors the logs for a kubectl port forward command
 // If it sees an error, it calls back to the EntryManager to
 // retry the entire port forward operation.
-func (*KubectlForwarder) Monitor(p *portForwardEntry, retryFunc func()) {
+func (*KubectlForwarder) monitorErrorLogs(ctx context.Context, buf *bytes.Buffer, cmd *exec.Cmd, p *portForwardEntry) {
 	for {
-		time.Sleep(1 * time.Second)
-		s, _ := p.logBuffer.ReadString(byte('\n'))
-		if s != "" {
-			logrus.Tracef("[port-forward] %s", s)
-			if strings.Contains(s, "error forwarding port") || strings.Contains(s, "unable to forward") {
-				// kubectl is having an error. retry the command
-				logrus.Infof("retrying kubectl port-forward due to error: %s", s)
-				go retryFunc()
-				return
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			time.Sleep(1 * time.Second)
+			s, _ := buf.ReadString(byte('\n'))
+			if s != "" {
+				logrus.Tracef("[port-forward] %s", s)
+
+				if strings.Contains(s, "error forwarding port") ||
+					strings.Contains(s, "unable to forward") ||
+					strings.Contains(s, "error upgrading connection") {
+					// kubectl is having an error. retry the command
+					logrus.Tracef("killing port forwarding %v", p)
+					if err := cmd.Process.Kill(); err != nil {
+						logrus.Tracef("failed to kill port forwarding %v, err: %s", p, err)
```

Comment:
Can we also print the pid for the process that did not get killed?
Users can kill it themselves.

Reply:
the typical case that I saw this error happen is when the process is already killed...it is really just a trace for the sake of completeness, I don't really think that the PID is necessary.